flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java

```diff
@@ -168,6 +175,12 @@ public void run() {
 	Optional<BufferAndAvailability> getNextBuffer() throws IOException, InterruptedException {
 		checkError();
 
+		BufferAndAvailability bufferAndAvailability = getNextRecoveredStateBuffer();
+		if (bufferAndAvailability != null) {
```

Comment:
My previous assumption was that the local channel will not be chosen by `SingleInputGate` to read if there were no buffers to insert into `RecoveredInputChannel#receivedBuffers` to notify `SingleInputGate.notifyChannelNonEmpty` before.
Or I missed some other corner case?

Reply:
Generally speaking there are no guarantees that data notifications will be accurate or not (I'm not sure if that's the case here). But the code looks a bit strange here and it relays on some extra assumptions.

Reply:
I know that the current data notification from downstream side is not very accurate which might cause return `null` buffer sometimes, but we can make the channel state notification accurate (actually it is the ground truth now). Then another truth is that the channel state consumption should always happen before downstream consumption, so it should be no problem here. 

As long as `getNextBuffer` is triggered by channel state buffer, it can always get accurate buffer. Otherwise if it is triggered by downstream side notification, then it also confirms that the channel state recovery should already finish before.